jobserver/views/users.py

```diff
@@ -1,82 +1,119 @@
+
+        if request.user.is_authenticated:
+            return redirect(self.next_url)
+
+        return super().dispatch(request, *args, **kwargs)
 
     def form_valid(self, form):
         user = User.objects.filter(email=form.cleaned_data["email"]).first()
 
-        # we have three possible states for the given email:
-        # * unknown user
-        # * GitHub user
-        # * email-only user
-        # we want to show the same outcome regardless of state so a bad actor
-        # can't work out work out who uses the service by email
         if user:
+            if not has_role(user, InteractiveReporter):
+                messages.error(
+                    self.request,
+                    "Only users who have signed up to Interactive can log in via email",
+                )
+                return redirect("login")
+
             if user.social_auth.exists():
-                # send GitHub users a link to the PSA entrypoint and explain
-                # that they don't have a password to reset
+                # we don't want to expose users email address to a bad actor
+                # via the login page form so we're emailing them with a link
+                # to the login page to click the right button.
+                # We can't email them with a link to the social auth entrypoint
+                # because it only accepts POSTs.
                 send_github_login_email(user)
             else:
-                send_reset_password_email(user)
-
-        messages.success(
-            self.request, "Your password reset request was successfully sent."
-        )
-        return redirect("/")
-
-
-class SetPassword(PasswordResetConfirmView):
-    post_reset_login = True
-    post_reset_login_backend = "django.contrib.auth.backends.ModelBackend"
-    success_url = "/"
-    template_name = "set_password.html"
-
-    def get_success_url(self):
-        # the authorization framework doesn't support getting all roles for a
-        # user across all objects, so we hae to manually do it here.
-        roles = set(self.request.user.roles)
-        for membership in self.request.user.project_memberships.all():
-            roles |= set(membership.roles)
-
-        # treat users with only the InteractiveReporter role as interactive users.
-        # we'll have users with this role along with others, eg for staff, but
-        # they likely don't want to be redirected on login.
-        if roles != {InteractiveReporter}:
-            return "/"
-
-        # if the user is logging in via this view, so with an email address and
-        # password, _and_ has the InteractiveReporter role then they should
-        # definitely be an interactive user.
-
-        # TODO: add a page for users who have 2+ projects
-        project = self.request.user.projects.first()
-        return project.get_interactive_url()
+                self.request.session[INTERNAL_LOGIN_SESSION_TOKEN] = user.email
```

Comment:
We need to store the token in the session to compare.

This means restucturing all the auth bits to be unrelated to the mode (which they are not really anymore)

e.g.

```
token = secrets.token_urlsafe(64)
self.request.session[INTERNAL_LOGIN_SESSION_TOKEN] = (user.email, token)
signed_token = TimestampSigner(salt="login").sign(token)
url = reverse("login-with-url", kwargs={"token": signed_token})
send_login_email(
    user, 
    url=url,
    timeout_minutes=settings.LOGIN_URL_TIMEOUT_MINUTES,
)
```